library/spdm_requester_lib/libspdm_req_encap_event_ack.c

```diff
@@ -0,0 +1,326 @@
+
+        if ((index != 0) && events_list_is_sequential) {
+            if (event_instance_id != (prev_event_instance_id + 1)) {
+                events_list_is_sequential = false;
+            }
+        }
+        if (event_instance_id < event_instance_id_min) {
+            event_instance_id_min = event_instance_id;
+        }
+        if (event_instance_id > event_instance_id_max) {
+            event_instance_id_max = event_instance_id;
+        }
+
+        event_instance_id_sum += event_instance_id;
+        prev_event_instance_id = event_instance_id;
+
+        ptr += sizeof(uint32_t) + sizeof(uint32_t);
+        svh_id = *ptr;
+        ptr += sizeof(uint8_t);
+        svh_vendor_id_len = *ptr;
+        ptr += sizeof(uint8_t);
+
+        if (!libspdm_validate_svh_vendor_id_len(svh_id, svh_vendor_id_len)) {
+            return libspdm_generate_encap_error_response(
+                context, SPDM_ERROR_CODE_INVALID_REQUEST, 0, response_size, response);
+        }
+
+        if (!check_for_space(ptr, end_ptr,
+                             (size_t)svh_vendor_id_len + sizeof(uint16_t) + sizeof(uint16_t))) {
+            return libspdm_generate_encap_error_response(
+                context, SPDM_ERROR_CODE_INVALID_REQUEST, 0, response_size, response);
+        }
+
+        ptr += svh_vendor_id_len;
+
+        event_type_id = libspdm_read_uint16(ptr);
+        ptr += sizeof(uint16_t);
+        event_detail_len = libspdm_read_uint16(ptr);
+        ptr += sizeof(uint16_t);
+
+        if (svh_id == SPDM_REGISTRY_ID_DMTF) {
+            if (!validate_dmtf_event_type(event_type_id, event_detail_len)) {
+                return libspdm_generate_encap_error_response(
+                    context, SPDM_ERROR_CODE_INVALID_REQUEST, 0, response_size, response);
+            }
+        }
+
+        if (!check_for_space(ptr, end_ptr, (size_t)event_detail_len)) {
+            return libspdm_generate_encap_error_response(
+                context, SPDM_ERROR_CODE_INVALID_REQUEST, 0, response_size, response);
+        }
+
+        ptr += event_detail_len;
+        calculated_request_size += sizeof(uint32_t) + sizeof(uint32_t) + sizeof(uint8_t) +
+                                   sizeof(uint8_t) + (size_t)svh_vendor_id_len + sizeof(uint16_t) +
+                                   sizeof(uint16_t) + (size_t)event_detail_len;
+    }
+
+    /* Event must be sent in a secure session so message size can be calculated exactly. */
+    if (request_size != calculated_request_size) {
+        return libspdm_generate_encap_error_response(
+            context, SPDM_ERROR_CODE_INVALID_REQUEST, 0, response_size, response);
+    }
+
+    /* Check that sum of event instance IDs is consistent with sequential event instance IDs.
+     * This check may pass if there are gaps or duplicates in event instance IDs.
+     * Those will be discovered when the events are sent to the Integrator. */
+    if (!events_list_is_sequential) {
+        if ((event_instance_id_sum * 2) !=
```

Comment:
@jyao1 recommends that this be replaced with a loop through `find_event_instance_id` to check that there are no gaps or duplicates. That way the Integrator will not get partial events from libspdm.